storage: never fuzz GC scores up

tbg opened the pull request:
I haven't seen this be an issue in practice, but it seems good to be prudent.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20716)
<!-- Reviewable:end -->

nvanbenschoten commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20716#-L0LQxqu6anvBmGy79of:-L0LQxqu6anvBmGy79og:b-csdkb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/577a9f02976c43315ce98c82c73703843d8c6d53/pkg/storage/gc_queue.go#L299)):*
> ```Go
> 	// and otherwise behaves indistinguishable given that we can't possibly hope
> 	// to GC values faster than that.
> 	if r.TTL == 0 {
> ```

Not changed here, but should this just be `r.TTL = max(r.TTL, time.Second)` (i.e. should anything under a second be rounded up)? Wouldn't this give a TTL of 0.5 seconds a higher priority that one with a TTL of zero seconds? 

---

*[pkg/storage/gc_queue.go, line 344 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20716#-L0LS5MfMeIoPuVI8CeT:-L0LS5MfMeIoPuVI8CeU:b-1a4bmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/577a9f02976c43315ce98c82c73703843d8c6d53/pkg/storage/gc_queue.go#L344)):*
> ```Go
> 	// normalizes. Note that at the time of writing this criterion hasn't
> 	// undergone a reality check yet.
> 	r.IntentScore = ms.AvgIntentAge(now.WallTime) / float64(intentAgeNormalization.Nanoseconds()/1E9)
> ```

Again, not changed here, but why `/1E9`? Doesn't the `.Nanoseconds()` already adjust the units to be the same? Or does `AvgIntentAge` return seconds and I just can't find any comments explaining that?

---

*[pkg/storage/gc_queue.go, line 346 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20716#-L0LTdOoc_UyTFoMFxaW:-L0LTdOoc_UyTFoMFxaX:b9pmg71) ([raw file](https://github.com/cockroachdb/cockroach/blob/577a9f02976c43315ce98c82c73703843d8c6d53/pkg/storage/gc_queue.go#L346)):*
> ```Go
> 	r.IntentScore = ms.AvgIntentAge(now.WallTime) / float64(intentAgeNormalization.Nanoseconds()/1E9)
> 
> 	// Random factor in [0.75, 1.00] to cause decoherence of replicas with
> ```

This will reduce the average `valScore` and `intentScore` by 12.5%, making the gcQueue less aggressive overall. Do we need to adjust the score thresholds to compensate?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20716)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: never fuzz GC scores up

I haven't seen this be an issue in practice, but it seems good to be prudent.

Release note: None

tschottdorf commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/gc_queue.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20716#-L0LQxqu6anvBmGy79of:-L0QLsMsFKKpcJj8Vqq1:b-4624g5) ([raw file](https://github.com/cockroachdb/cockroach/blob/577a9f02976c43315ce98c82c73703843d8c6d53/pkg/storage/gc_queue.go#L299)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Not changed here, but should this just be `r.TTL = max(r.TTL, time.Second)` (i.e. should anything under a second be rounded up)? Wouldn't this give a TTL of 0.5 seconds a higher priority that one with a TTL of zero seconds? 
</blockquote></details>

TTL is a multiple of `time.Second`, but I've made this `r.TTL <= time.Second` to make it look less confusing for future readers.

---

*[pkg/storage/gc_queue.go, line 344 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20716#-L0LS5MfMeIoPuVI8CeT:-L0QMEoWskEf9kFb2q34:b4ostl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/577a9f02976c43315ce98c82c73703843d8c6d53/pkg/storage/gc_queue.go#L344)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Again, not changed here, but why `/1E9`? Doesn't the `.Nanoseconds()` already adjust the units to be the same? Or does `AvgIntentAge` return seconds and I just can't find any comments explaining that?
</blockquote></details>

The age is measured in seconds. For example, if an intent is around from `t=2.5s` to `t=4.1s` (the current time) then it contributes an intent age of two seconds (one second picked up when crossing `t=3s`, another one at `t=4s`). I added a comment.

---

*[pkg/storage/gc_queue.go, line 346 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20716#-L0LTdOoc_UyTFoMFxaW:-L0QMzYDxpB_YI-ecUGn:b-bh879o) ([raw file](https://github.com/cockroachdb/cockroach/blob/577a9f02976c43315ce98c82c73703843d8c6d53/pkg/storage/gc_queue.go#L346)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This will reduce the average `valScore` and `intentScore` by 12.5%, making the gcQueue less aggressive overall. Do we need to adjust the score thresholds to compensate?
</blockquote></details>

That's a good point. Made this `[0.95,1]`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20716)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 4 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20716#-:-L0T41zMwj2ZuJv7_EKN:bnfp4nl)*
<!-- Sent from Reviewable.io -->
